### PR TITLE
[Common][PyTorch]Current Scaling Group Quantization + Enabling Varying Last/Both Dims in Group Quantize

### LINKED ISSUE #2449: Quantization support for GroupedTensor: FP8 per-tensor
Implement quantization support for the GroupedTensor type with FP8 per-tensor quantization.
The needed modifications to the existing kernel:
 - handling multiple amax values, each for different tensor
 - ignore padding in the allocation

### PR DESCRIPTION
**Pull request by vthumbe1503**
# Description
- Current Scaling Group Quantization added in Common/Pytorch for Varying First/Last Dims
- Enables Varying Last/Both Dims for Group Quantize from Pytorch C++/Python Infrastructure(Wasnt present before for MXFP8 group quantize)
- Enables No_op flag from group_quantize Pytorch infra.(This was present previously only in Pytorch single tensor quantization infra)
- Benchmarking Scripts/Tests to benchmark Group Quantize Current Scaling for corner cases like overallocation, non-uniform group sizes etc. Tests are added for both Common(CPP) and Pytorch(Python)

Performance from Benchmarking Script for Current Scaling Group Quantize.
 _NOTE:_ 
1. overalloc means the grouped tensor is overallocated 4 times the actual size. But this doesnt matter to the kernel since kernel reads the offsets from the device and decides the CTA to elements mapping. We test this because this is a valid cuda graph use-case.
2. zipf, heavy, mild etc represent the amount of group imbalance. As seen below, the kernel doesnt deteriorate in performance if group imbalance is present in Blackwell. For Hopper, group_imbalance affects the perf of columnwise cast kernel at the moment due to the use of 2D grid. This will be fixed in a followup PR.

**GB200**
image

**H100**
image
image
image

Fixes #2449 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- **Core Changes**
  - `common.cuh` has been seperated out into two files  -- `grouped_layout.cuh` and `grouped_tma.cuh`. The idea is that, `grouped_layout.cuh` has generic utilities commonly used in grouped tensor kernels. And `grouped_tma.cuh` has the arch specific TMA changes. This is done so that current scaling can still be non-arch specific and it can use just the `grouped_layout.cuh` since we didnt need TMA for current scaling.
  - _Dispatch_ has the change to enable calling group quantize for NVTE_DELAYED_SCALING which is the enum used for both delayed scaling and current scaling in the legacy infra(**Note**: Delayed Scaling hasnt enabled in the Pytorch Infra in this PR)
  - _FP8_ : Kernels for Computing Grouped Amax and grouped Cast to Per tensor FP8 are added. Both hit close to 6TB/s to all use-cases on Blackwell and 3 TB/s for H100 rowwise cast kernel and 2.5 TB/s for H100 columnwise cast kernel.
- **Other Common Changes**
  - NVTE APIs for computing grouped amax and computing grouped scale from grouped amax added which uses the kernels implemented in core.
  - nvte_splits_to_offsets_2d --> NVTE API for computing for computing offsets for VARYING_BOTH_DIMS added. Since we now enable enabling varying last/both dims from pytorch infra.

- **Pytorch Group Quantize API Changes**
  - Current Scaling Support added that uses the NVTE APIs from common for computing grouped amax and performing a grouped cast
  - Accepts last_dims in the API that enables Varying Last Dims/Both Dims in the Pytorch infrastructure
  - Accepts no_op flag for cuda graph when wanting skip_fp8_weight_update. (Would be needed for single_grouped_weight correctness with cuda graph when skip_weight_update is enabled --> essentially when we use group_quantize on the weights)
 
- **Benchmarking Scripts/Tests added for Current Scaling** for Varying First/Last Dims and to handle overallocation of grouped tensors. Varying All Dims is not supported yet.

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+constexpr int GROUPED_AMAX_KERNEL_THREADS = 256;
+// Target block-per-tensor multiplier so that, even with few groups, we cover
+// enough SMs on H100/GB200-class GPUs (132 SMs) to be HBM-bandwidth bound.
+// blocks_per_tensor is bounded by both work and a hard cap to avoid launching
+// far more blocks than there is work for.
+constexpr int GROUPED_AMAX_BLOCKS_PER_TENSOR_CAP = 64;
+constexpr size_t GROUPED_AMAX_MIN_ELTS_PER_BLOCK = 8 * 1024;  // ~16KB of bf16
+
+// Zero per-tensor amax buffer so the main kernel can use atomicMax updates.
+__launch_bounds__(GROUPED_AMAX_KERNEL_THREADS) __global__
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: Alternatively, we can clear the contiguous buffer with cudaMemsetAsync() on the same stream before launching the main kernel. This preserves ordering and may avoid the overhead of a separate zeroing kernel.

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      if (aligned) {
+        InputType acc0 = 0.f;
+        InputType acc1 = 0.f;
+        InputType acc2 = 0.f;
+        InputType acc3 = 0.f;
+        size_t v = tid;
+        // 4-way vectorized load and reduce
+        for (; v + 3 * blockDim.x < total_vecs; v += 4 * blockDim.x) {
+          IVecT vec0, vec1, vec2, vec3;
+          vec0.load_from(base + v * NVEC);
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_amax_fp8.cuh:146]: Let’s benchmark the LDG path here. If it turns out to be a bottleneck, we may need to switch to TMA.

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        thread_amax = fmaxf(thread_amax, __shfl_xor_sync(0xFFFFFFFFu, thread_amax, s));
+      }
+
+      const int warp_id = tid / THREADS_PER_WARP;
+      const int lane = tid % THREADS_PER_WARP;
+      if (lane == 0) {
+        s_block_amax[warp_id] = thread_amax;
+      }
+      __syncthreads();
+      // Reduce amax within the warp
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: We can reuse the warp-level reducer from `transformer_engine/common/utils.cuh` here to avoid duplicating the reduction logic.

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+      for (size_t i = tail_start + tid; i < numel; i += blockDim.x) {
+        thread_amax_val = max_val(thread_amax_val, abs_val(base[i]));
+      }
+
+      // Warp-shuffle reduce within the block
+      float thread_amax = static_cast<float>(thread_amax_val);
+#pragma unroll
+      for (int s = THREADS_PER_WARP / 2; s > 0; s >>= 1) {
+        thread_amax = fmaxf(thread_amax, __shfl_xor_sync(0xFFFFFFFFu, thread_amax, s));
+      }
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: We can also reuse the warp-level reducer here.

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        // 4-way vectorized load and reduce
+        for (; v + 3 * blockDim.x < total_vecs; v += 4 * blockDim.x) {
+          IVecT vec0, vec1, vec2, vec3;
+          vec0.load_from(base + v * NVEC);
+          vec1.load_from(base + (v + blockDim.x) * NVEC);
+          vec2.load_from(base + (v + 2 * blockDim.x) * NVEC);
+          vec3.load_from(base + (v + 3 * blockDim.x) * NVEC);
+#pragma unroll
+          for (int i = 0; i < NVEC; ++i) {
+            acc0 = max_val(acc0, abs_val(vec0.data.elt[i]));
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_amax_fp8.cuh:152]: If this becomes performance-critical, we could consider reusing helper functions from `transformer_engine/common/util/ptx.cuh`, such as abs_max_2x(), or adding a custom inline PTX helper if the existing ones do not fit.

File: transformer_engine/common/cast/fp8/group_quantize_fp8.cuh [hunk 6]
```diff
@@ -0,0 +1,1350 @@
...
+template <typename IType, typename OType>
+struct supports_fast_scaled_fp8_cvt_4 {
+  static constexpr bool value =
+      (std::is_same<IType, bf16>::value || std::is_same<IType, fp16>::value ||
+       std::is_same<IType, fp32>::value) &&
+      (std::is_same<OType, fp8e4m3>::value || std::is_same<OType, fp8e5m2>::value);
+};
+
+template <typename IType, typename OType>
+__device__ __forceinline__ void fast_scaled_fp8_cvt_4(const IType *input, OType *output,
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_quantize_fp8.cuh:54]: Could we rewrite this function to reduce code duplication and hardcoding? ptx::mul_cvt_4x is overloaded, so we should be able to use it directly and let the types select the correct overload. If `input` and `output` are properly aligned for the corresponding `ptx::FPx4<>` types, we can reinterpret them as 4-element packed values and avoid duplicating the conversion logic, e.g.

```
using IType4 = ptx::FPx4<IType>;
using OType4 = ptx::FPx4<OType>;

const IType4& in4x = *reinterpret_cast<const IType4*>(input);
OType4& out4x = *reinterpret_cast<OType4*>(output);

ptx::mul_cvt_4x(out4x, in4x, scale_2x);
```

File: transformer_engine/common/cast/fp8/group_quantize_fp8.cuh [hunk 7]
```diff
@@ -0,0 +1,1350 @@
...
+                                reinterpret_cast<OType *>(output->data.dptr), scale_ptr, noop_ptr,
+                                num_tensors, total_elements, offsets_ptr);
+                      }
+                      launched_fast_path = true;
+                    }
+                  }
+                  // Varying-first columnwise intentionally falls through to the
+                  // generic group_cast_fp8_kernel below. The dedicated
+                  // group_cast_fp8_varying_first_tile_kernel used a grid-stride
+                  // row loop that spilled to 118 registers/thread (2 blocks/SM,
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_quantize_fp8.cuh]: Could we revise this comment to avoid relying on machine-specific benchmark data? The current numbers are tied to a specific machine/build configuration and may not generalize.

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh (hunk 2 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_amax_fp8.cuh:146]: LSU hits only at around 8.7%. We are consitentyl hitting 78-80% of bW on Blackwell with the kernel with this vectorization + 4 way unrolling, for all use-cases (same-shape, variable shape, overallocation, etc)

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh (hunk 5 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_amax_fp8.cuh:152]: We are memory bandwidth bound completely.
Long Scoreboard is 85.6%

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+  using namespace group_amax_kernel;
+  if (num_tensors == 0) {
+    return;
+  }
+  (void)shape_rep;
+  (void)first_dims_ptr;
+  (void)last_dims_ptr;
+
+  // Zero out the per-tensor amax buffer so atomicMaxFloat works.
+  const size_t zero_blocks =
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: Let’s reuse the existing `DIVUP()` helper from `transformer_engine/common/common.h`.

File: transformer_engine/pytorch/csrc/extensions/cast.cpp [hunk 9]
```diff
@@ -204,11 +207,100 @@ void group_quantize_nvfp4_impl(const GroupedTensorWrapper &grouped_input_tensor,
   });
 }
 
+float fp8_max_for_dtype(const DType dtype) {
+  switch (dtype) {
+    case DType::kFloat8E4M3:
```
Comment (Oleg-Goncharov) [inline transformer_engine/pytorch/csrc/extensions/cast.cpp]: Could we rewrite this function to reuse the existing constants (in `transformer_engine/common/common.h`) instead of redefining it? For example:

```
if (!is_fp8_dtype(dtype)) {
    NVTE_ERROR("Expected FP8 dtype for grouped current-scaling quantization, got ",
               to_string(dtype), ".");
}
TRANSFORMER_ENGINE_TYPE_SWITCH_FP8ONLY(dtype, T,
{
    return TypeExtrema<T>::max;
});
```

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh (hunk 3 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: Done

File: transformer_engine/common/cast/fp8/group_quantize_fp8.cuh (hunk 6 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_quantize_fp8.cuh:54]: Done

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh (hunk 1 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: zero kernel is used since we want no_op_ptr compatibility. I believe we want that in case of pipeline parallelism where we want to skip the quantization after first microbatch when using cuda graphs

We are doing this even for per tensor scaling in [here](https://github.com/NVIDIA/TransformerEngine/blob/318dd94a6d2cc492c838343712c0fef67ab6c2f0/transformer_engine/common/recipe/current_scaling.cu#L23-L24)

File: transformer_engine/pytorch/csrc/extensions/cast.cpp (hunk 9 above)
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/extensions/cast.cpp]: Done. Thanks!

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh (hunk 8 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: Done

File: transformer_engine/common/cast/fp8/group_quantize_fp8.cuh (hunk 7 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_quantize_fp8.cuh]: Revised the comment

File: transformer_engine/common/cast/fp8/group_amax_fp8.cuh (hunk 4 above)
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_amax_fp8.cuh]: Done

File: transformer_engine/common/cast/fp8/group_quantize_fp8.cuh [hunk 10]
```diff
@@ -0,0 +1,1226 @@
...
+    NVTE_CHECK(first_logical_dim % num_tensors == 0,
+               "First logical dimension must be divisible by num_tensors.");
+    const size_t rows_per_tensor = first_logical_dim / num_tensors;
+    work_blocks_Y = num_tensors * DIVUP(rows_per_tensor, row_tile_size);
+  } else if (shape_rep == ShapeRepresentation::VARYING_FIRST_DIM) {
+    // first_dims live on device; over-allocate the grid enough to cover one partial tile per group.
+    work_blocks_Y += num_tensors;
+  } else if (shape_rep == ShapeRepresentation::VARYING_LAST_DIM) {
+    // last_dims live on device; map Y to per-group row tiles and over-allocate X by total width.
+    work_blocks_Y = num_tensors * DIVUP(first_logical_dim, row_tile_size);
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/common/cast/fp8/group_quantize_fp8.cuh:1063]: decode_job is not used for this kernel.. that is used for the mxfp8 kernel. This kernel has 2D grid assigned and not a flat grid

**Comment by vthumbe1503** [comment]
Pipeline: 54747206

File: transformer_engine/pytorch/tensor/storage/grouped_tensor_storage.py [hunk 11]
```diff
@@ -693,9 +696,27 @@ def make_grouped_tensor(
         logical_shape = (logical_first_dim, logical_last_dim)
 
         no_quantization = quantizer is None
-
         rowwise_usage = quantizer.rowwise_usage if not no_quantization else True
         columnwise_usage = quantizer.columnwise_usage if not no_quantization else False
+        compatible_recipe = None if no_quantization else quantizer._get_compatible_recipe()
+
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/tensor/storage/grouped_tensor_storage.py:702]: Fixed the comment

File: transformer_engine/pytorch/csrc/extensions/cast.cpp [hunk 12]
```diff
@@ -204,17 +208,113 @@ void group_quantize_nvfp4_impl(const GroupedTensorWrapper &grouped_input_tensor,
...
+  std::array<NVTETensor *, 3> tensor_lists_ptr = {amax_list.data(), scale_list.data(),
+                                                  scale_inv_list.data()};
+
+  TensorWrapper noop_flag_cpp;
+  NVTE_SCOPED_GIL_RELEASE({
+    nvte_multi_tensor_compute_scale_and_scale_inv_cuda(
+        2048 * 32, noop_flag_cpp.data(), tensor_lists_ptr.data(), tensor_lists_ptr.size(),
+        /*num_tensors_per_list=*/1, fp8_max_for_dtype(quantizer_cpp->dtype),
+        quantizer_cpp->force_pow_2_scales, quantizer_cpp->amax_epsilon, stream);
+  });
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/extensions/cast.cpp]: We have the grouped impl used now for computing scale and scale inv

File: transformer_engine/pytorch/csrc/extensions/pybind.cpp [hunk 13]
```diff
@@ -169,7 +169,7 @@ void bind_quantize_with_amax_extensions(py::module_ &m) {
         py::arg("quantizer"), py::arg("rowwise_amax"), py::arg("columnwise_amax"));
   m.def("nvfp4_group_quantize_with_amax", nvfp4_group_quantize_with_amax, py::arg("tensor"),
         py::arg("quantizer"), py::arg("num_tensors"), py::arg("first_dims"),
-        py::arg("rowwise_amax"), py::arg("columnwise_amax"),
+        py::arg("last_dims") = py::none(), py::arg("rowwise_amax"), py::arg("columnwise_amax"),
         py::arg("tensor_offsets") = py::none());
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/extensions/pybind.cpp:173]: Fixed grouped_mlp callsite

**Review by Oleg-Goncharov — CHANGES_REQUESTED** [review]
Hi @vthumbe1503, could you please also add C++ unit tests covering various shapes, similar to grouped MXFP8?

File: transformer_engine/pytorch/csrc/quantizer.cpp [hunk 14]
```diff
@@ -713,23 +761,31 @@ std::pair<GroupedTensorWrapper, py::object> Float8CurrentScalingQuantizer::creat
   std::optional<at::Tensor> columnwise_scale_inv;
   at::Tensor scale = at::empty({static_cast<int64_t>(num_tensors)}, float_opts);
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/quantizer.cpp:776]: There is no rowwise_amax attribute. And we are already adding "amax" in kwargs.

File: transformer_engine/pytorch/csrc/extensions/cast.cpp [hunk 15]
```diff
@@ -276,7 +373,9 @@ py::object group_quantize(const at::Tensor &tensor, py::handle quantizer, const
     }
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/extensions/cast.cpp:373]: With this PR mxfp8 group quantize is supported for all the three cases, varying first, varying last and varying both dims. So, the guard isnt needed.

[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/extensions/cast.cpp:373]: Refer to the passing tests for mxfp8 group quantize in test_grouped_tensor.py

**Review by denera — CHANGES_REQUESTED** [review]
No new file-specific issues that haven't already been discussed, but it looks to me like there is no validation of `first_dims` / `last_dims` for `precomputed_tensor_offsets`. The validation is only present in `build_grouped_tensor_offsets()`, which is naturally skipped when we invoke `create_grouped_tensor()` with an already existing `precomputed_tensor_offsets`. So I would recommend maybe add a shared validator for tensor offsets that we can re-use in both code paths (precomputed and on-demand).

**Review by Oleg-Goncharov — APPROVED** [review]
LGTM

**Review by denera — APPROVED** [review]
LGTM!

**Comment by vthumbe1503** [comment]
Pipeline: 55606348
Errors are pre-existing on main
